pallets/treasury-buyout-extension/src/lib.rs

```diff
@@ -0,0 +1,525 @@
+#![deny(warnings)]
+#![cfg_attr(test, feature(proc_macro_hygiene))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[cfg(test)]
+mod mock;
+
+pub mod default_weights;
+
+#[cfg(test)]
+mod tests;
+
+mod types;
+
+use crate::{
+	default_weights::WeightInfo,
+	types::{AccountIdOf, Amount, BalanceOf, CurrencyIdOf},
+};
+use codec::{Decode, Encode};
+use frame_support::{
+	dispatch::{DispatchError, DispatchResult},
+	ensure,
+	sp_runtime::SaturatedConversion,
+	traits::{Get, IsSubType},
+};
+use orml_traits::MultiCurrency;
+pub use pallet::*;
+use sp_arithmetic::per_things::Rounding;
+use sp_runtime::{
+	traits::{DispatchInfoOf, One, SignedExtension, Zero},
+	transaction_validity::{
+		InvalidTransaction, TransactionValidity, TransactionValidityError, ValidTransaction,
+	},
+	ArithmeticError, FixedPointNumber, FixedU128,
+};
+use sp_std::{fmt::Debug, marker::PhantomData};
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::{pallet_prelude::*, sp_runtime::Permill};
+	use frame_system::{ensure_root, ensure_signed, pallet_prelude::*};
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + orml_currencies::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Used for currency-related operations
+		type Currency: MultiCurrency<AccountIdOf<Self>, CurrencyId = CurrencyIdOf<Self>>;
+
+		/// Used for getting the treasury account
+		#[pallet::constant]
+		type TreasuryAccount: Get<AccountIdOf<Self>>;
+
+		/// Buyout period in blocks
+		#[pallet::constant]
+		type BuyoutPeriod: Get<u32>;
+
+		/// Fee from the native asset buyouts
+		#[pallet::constant]
+		type SellFee: Get<Permill>;
+
+		/// Type that allows for checking if currency type is ownable by users
+		type AllowedCurrencyIdVerifier: AllowedCurrencyIdVerifier<CurrencyIdOf<Self>>;
```

Comment:
The name is a bit clunky. Is `AllowedCurrencyChecker` better? Not sure

Reply:
Yes, indeed it's a bit clunky and your suggestion sounds good. I wanted this name to not be easily confused with `CurrencyIdChecker` from the `orml-tokens-management-extension` pallet